File: tools/wiki.py
#!/usr/bin/env python3
"""wiki.py — deterministic tooling for an LLM wiki.

The LLM compiles raw/ into wiki/. This script checks the build.
No API key, no network, no LLM: it verifies mechanical invariants only.
It cannot tell you whether a page is true.

    lint     check every invariant; exit 1 on errors
    status   what needs recompiling (make-style dependency check)
    index    regenerate wiki/index.md from frontmatter
    graph    emit the link graph as mermaid
    stats    counts and coverage
    new      scaffold a page with correct frontmatter
"""
from __future__ import annotations

import argparse
import datetime as dt
import re
import sys
from collections import defaultdict
from dataclasses import dataclass, field
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
RAW = ROOT / "raw"
WIKI = ROOT / "wiki"

PAGE_TYPES = {"concept", "entity", "comparison", "question", "overview"}
STATUSES = {"established", "contested", "provisional", "superseded"}
REQUIRED = ("title", "type", "status", "updated", "sources")
# Pages exempt from the orphan check and from body-level citation rules.
INFRA = {"index", "log", "synthesis"}
SPECIAL_CITES = {"synthesis"}

LINK_RE = re.compile(r"\[\[([^\]|]+?)(?:\|[^\]]*)?\]\]")
CITE_RE = re.compile(r"\[\^([A-Za-z0-9._-]+)\]")
FOOTNOTE_DEF_RE = re.compile(r"^\[\^([A-Za-z0-9._-]+)\]:")
FENCE_RE = re.compile(r"^\s*```")


# ---------------------------------------------------------------- frontmatter

def parse_frontmatter(text: str) -> tuple[dict, str]:
    """Parse the YAML subset we actually use: scalars and '- ' lists."""
    if not text.startswith("---"):
        return {}, text
    end = text.find("\n---", 3)
    if end == -1:
        return {}, text
    block = text[3:end].strip("\n")
    body = text[end + 4:].lstrip("\n")

    meta: dict = {}
    key = None
    for line in block.splitlines():
        if not line.strip() or line.lstrip().startswith("#"):
            continue
        if line.startswith((" ", "\t")) and line.lstrip().startswith("- ") and key:
            meta.setdefault(key, [])
            if isinstance(meta[key], list):
                meta[key].append(line.lstrip()[2:].strip().strip("\"'"))
            continue
        if ":" in line:
            key, _, val = line.partition(":")
            key = key.strip()
            val = val.strip()
            if val.startswith("[") and val.endswith("]"):
                # inline list: [] or [a, b]
                inner = val[1:-1].strip()
                meta[key] = [i.strip().strip("\"'")
                             for i in inner.split(",") if i.strip()]
            else:
                val = val.strip("\"'")
                meta[key] = val if val else []
    return meta, body


def strip_code(body: str) -> str:
    """Remove fenced code blocks so examples in docs don't trip the checks."""
    out, fenced = [], False
    for line in body.splitlines():
        if FENCE_RE.match(line):
            fenced = not fenced
            continue
        if not fenced:
            out.append(line)
    return "\n".join(out)


def as_date(val) -> dt.date | None:
    if not isinstance(val, str):
        return None
    for fmt in ("%Y-%m-%d", "%Y-%m", "%Y"):
        try:
            return dt.datetime.strptime(val.strip(), fmt).date()
        except ValueError:
            continue
    return None


# ---------------------------------------------------------------------- model

@dataclass
class Page:
    slug: str
    path: Path
    meta: dict
    body: str
    links: list[str] = field(default_factory=list)
    cites: list[str] = field(default_factory=list)

    @property
    def is_infra(self) -> bool:
        return self.slug in INFRA


@dataclass
class Source:
    sid: str
    path: Path
    meta: dict


def load() -> tuple[dict[str, Page], dict[str, Source]]:
    pages: dict[str, Page] = {}
    for p in sorted(WIKI.rglob("*.md")):
        text = p.read_text(encoding="utf-8")
        meta, body = parse_frontmatter(text)
        slug = p.relative_to(WIKI).with_suffix("").as_posix()
        clean = strip_code(body)
        # Citations, ignoring footnote-definition lines at the page foot.
        cites = []
        for line in clean.splitlines():
            if FOOTNOTE_DEF_RE.match(line.strip()):
                continue
            cites.extend(CITE_RE.findall(line))
        pages[slug] = Page(slug, p, meta, body, LINK_RE.findall(clean), cites)

    sources: dict[str, Source] = {}
    for p in sorted(RAW.glob("*.md")):
        meta, _ = parse_frontmatter(p.read_text(encoding="utf-8"))
        sid = meta.get("source_id")
        if isinstance(sid, str) and sid:
            sources[sid] = Source(sid, p, meta)
    return pages, sources


# ----------------------------------------------------------------------- lint

class Report:
    def __init__(self) -> None:
        self.errors: list[str] = []
        self.warnings: list[str] = []

    def err(self, where: str, msg: str) -> None:
        self.errors.append(f"{where}: {msg}")

    def warn(self, where: str, msg: str) -> None:
        self.warnings.append(f"{where}: {msg}")


def lint(pages: dict[str, Page], sources: dict[str, Source]) -> Report:
    r = Report()
    inbound: dict[str, set[str]] = defaultdict(set)

    for slug, pg in sorted(pages.items()):
        where = pg.path.relative_to(ROOT).as_posix()

        # -- frontmatter ------------------------------------------------
        if not pg.meta:
            r.err(where, "no frontmatter")
            continue
        for fld in REQUIRED:
            if fld not in pg.meta:
                r.err(where, f"frontmatter missing '{fld}'")
        ptype = pg.meta.get("type")
        if ptype and ptype not in PAGE_TYPES:
            r.err(where, f"type '{ptype}' not in {sorted(PAGE_TYPES)}")
        status = pg.meta.get("status")
        if status and status not in STATUSES:
            r.err(where, f"status '{status}' not in {sorted(STATUSES)}")
        if "updated" in pg.meta and as_date(pg.meta["updated"]) is None:
            r.err(where, f"updated '{pg.meta['updated']}' is not a date")

        declared = pg.meta.get("sources") or []
        if isinstance(declared, str):
            declared = [declared] if declared else []

        # -- links ------------------------------------------------------
        for link in pg.links:
            target = link.strip().lstrip("/")
            if target not in pages:
                r.err(where, f"broken link [[{target}]]")
            else:
                inbound[target].add(slug)

        # -- citations --------------------------------------------------
        for cid in set(pg.cites):
            if cid in SPECIAL_CITES:
                continue
            if cid not in sources:
                r.err(where, f"citation [^{cid}] has no source in raw/")
            elif cid not in declared:
                r.err(where, f"cites [^{cid}] but omits it from frontmatter sources")
        for sid in declared:
            if sid not in sources:
                r.err(where, f"frontmatter lists unknown source '{sid}'")
            elif sid not in pg.cites and not pg.is_infra:
                r.warn(where, f"declares source '{sid}' but never cites it")

        # -- status obligations -----------------------------------------
        clean = strip_code(pg.body)
        if status == "contested" and not re.search(r"^##+\s*Contradictions",
                                                   clean, re.M | re.I):
            r.err(where, "status 'contested' requires a '## Contradictions' section")
        if status == "superseded" and not pg.links:
            r.err(where, "status 'superseded' must link forward to its replacement")

        # -- staleness (make-style dependency check) --------------------
        pdate = as_date(pg.meta.get("updated", ""))
        if pdate:
            for sid in declared:
                src = sources.get(sid)
                if not src:
                    continue
                sdate = as_date(src.meta.get("added", ""))
                if sdate and sdate > pdate:
                    r.err(where, f"stale: source '{sid}' added {sdate} "
                                 f"but page compiled {pdate} — recompile")

        # -- uncited assertions (heuristic, warning only) ----------------
        if not pg.is_infra:
            for para in re.split(r"\n\s*\n", clean):
                p = para.strip()
                if (not p or p.startswith(("#", ">", "|", "-", "*", "1.", "["))
                        or len(p.split()) < 25):
                    continue
                if not CITE_RE.search(p):
                    r.warn(where, f"uncited assertion: \"{' '.join(p.split())[:60]}…\"")

    # -- orphans --------------------------------------------------------
    for slug, pg in sorted(pages.items()):
        if pg.is_infra or not pg.meta:
            continue
        if not inbound.get(slug):
            r.warn(pg.path.relative_to(ROOT).as_posix(),
                   "orphan — no page links here")
        elif len(pg.links) < 3:
            r.warn(pg.path.relative_to(ROOT).as_posix(),
                   f"only {len(pg.links)} outbound links (aim for 3+)")

    # -- uningested sources ---------------------------------------------
    cited = {c for pg in pages.values() for c in pg.cites}
    for sid, src in sorted(sources.items()):
        if sid not in cited:
            r.err(src.path.relative_to(ROOT).as_posix(),
                  f"source '{sid}' is in raw/ but no wiki page cites it — not ingested")
    return r


def cmd_lint(args) -> int:
    pages, sources = load()
    r = lint(pages, sources)
    for w in r.warnings:
        print(f"  warn  {w}")
    for e in r.errors:
        print(f"  ERROR {e}")
    print()
    print(f"{len(pages)} pages, {len(sources)} sources, "
          f"{len(r.errors)} errors, {len(r.warnings)} warnings")
    if r.errors:
        print("\nBuild is broken. Fix the errors, then re-run.")
        return 1
    print("Wiki is consistent. (Consistent is not the same as true.)")
    return 0


def cmd_status(args) -> int:
    """make-style: which pages are older than the sources they depend on."""
    pages, sources = load()
    stale, uningested = [], []
    cited = {c for pg in pages.values() for c in pg.cites}
    for slug, pg in sorted(pages.items()):
        pdate = as_date(pg.meta.get("updated", ""))
        declared = pg.meta.get("sources") or []
        if isinstance(declared, str):
            declared = [declared]
        for sid in declared:
            src = sources.get(sid)
            sdate = as_date(src.meta.get("added", "")) if src else None
            if pdate and sdate and sdate > pdate:
                stale.append((slug, sid, pdate, sdate))
    for sid, src in sorted(sources.items()):
        if sid not in cited:
            uningested.append(sid)

    if not stale and not uningested:
        print("Up to date — every source is compiled, every page current.")
        return 0
    if uningested:
        print("Never ingested:")
        for sid in uningested:
            print(f"  {sid}  ({sources[sid].path.relative_to(ROOT)})")
    if stale:
        print("Stale (source newer than page):")
        for slug, sid, pdate, sdate in stale:
            print(f"  {slug}  depends on {sid} ({sdate}) but compiled {pdate}")
    print("\nRun the compile operation on these pages.")
    return 0


def cmd_index(args) -> int:
    pages, sources = load()
    by_type: dict[str, list[Page]] = defaultdict(list)
    for slug, pg in sorted(pages.items()):
        if pg.is_infra or not pg.meta:
            continue
        by_type[str(pg.meta.get("type", "unknown"))].append(pg)

    today = dt.date.today().isoformat()
    out = [
        "---",
        "title: Index",
        "type: overview",
        "status: established",
        f"updated: {today}",
        "sources: []",
        "---",
        "",
        "# Index",
        "",
        f"*Generated by `tools/wiki.py index` on {today}. Do not hand-edit.*",
        "",
        f"{len(pages)} pages compiled from {len(sources)} sources in `raw/`.",
        "",
    ]
    for ptype in ("concept", "entity", "comparison", "question", "overview"):
        group = by_type.get(ptype)
        if not group:
            continue
        out.append(f"## {ptype.capitalize()}s")
        out.append("")
        out.append("| page | status | updated | sources |")
        out.append("|---|---|---|---|")
        for pg in group:
            srcs = pg.meta.get("sources") or []
            if isinstance(srcs, str):
                srcs = [srcs]
            out.append(f"| [[{pg.slug}]] | `{pg.meta.get('status','?')}` "
                       f"| {pg.meta.get('updated','?')} | {len(srcs)} |")
        out.append("")

    out += ["## Sources", "", "| id | title | added | capture |", "|---|---|---|---|"]
    for sid, src in sorted(sources.items()):
        out.append(f"| `{sid}` | {src.meta.get('title','?')} "
                   f"| {src.meta.get('added','?')} | {src.meta.get('capture','?')} |")
    out.append("")

    (WIKI / "index.md").write_text("\n".join(out), encoding="utf-8")
    print(f"Wrote wiki/index.md — {len(pages)} pages, {len(sources)} sources.")
    return 0


def cmd_graph(args) -> int:
    pages, _ = load()
    print("```mermaid")
    print("graph LR")
    for slug, pg in sorted(pages.items()):
        if pg.is_infra:
            continue
        node = slug.replace("/", "_").replace("-", "_")
        label = pg.meta.get("title", slug)
        print(f'  {node}["{label}"]')
    for slug, pg in sorted(pages.items()):
        if pg.is_infra:
            continue
        a = slug.replace("/", "_").replace("-", "_")
        for link in sorted(set(pg.links)):
            if link in pages and link not in INFRA:
                b = link.replace("/", "_").replace("-", "_")
                print(f"  {a} --> {b}")
    print("```")
    return 0


def cmd_stats(args) -> int:
    pages, sources = load()
    types: dict[str, int] = defaultdict(int)
    statuses: dict[str, int] = defaultdict(int)
    words = links = cites = 0
    for pg in pages.values():
        types[str(pg.meta.get("type", "?"))] += 1
        statuses[str(pg.meta.get("status", "?"))] += 1
        words += len(strip_code(pg.body).split())
        links += len(pg.links)
        cites += len(pg.cites)
    raw_words = sum(len(p.read_text(encoding="utf-8").split())
                    for p in RAW.glob("*.md"))
    print(f"pages       {len(pages)}")
    print(f"sources     {len(sources)}")
    print(f"raw words   {raw_words}")
    print(f"wiki words  {words}  ({words / raw_words:.2f}x raw)"
          if raw_words else f"wiki words  {words}")
    print(f"links       {links}  ({links / max(len(pages),1):.1f} per page)")
    print(f"citations   {cites}  ({cites / max(len(pages),1):.1f} per page)")
    print("types       " + ", ".join(f"{k}={v}" for k, v in sorted(types.items())))
    print("status      " + ", ".join(f"{k}={v}" for k, v in sorted(statuses.items())))
    return 0


def cmd_new(args) -> int:
    slug = args.slug.strip("/").removesuffix(".md")
    path = WIKI / f"{slug}.md"
    if path.exists():
        print(f"{path.relative_to(ROOT)} already exists", file=sys.stderr)
        return 1
    path.parent.mkdir(parents=True, exist_ok=True)
    title = args.title or slug.rsplit("/", 1)[-1].replace("-", " ").capitalize()
    path.write_text(
        "---\n"
        f"title: {title}\n"
        f"type: {args.type}\n"
        f"status: {args.status}\n"
        f"updated: {dt.date.today().isoformat()}\n"
        "sources: []\n"
        "---\n\n"
        f"# {title}\n\n"
        "<!-- Every claim needs a [^source-id]. Link 3+ related pages. -->\n",
        encoding="utf-8")
    print(f"Created {path.relative_to(ROOT)}")
    return 0


def main() -> int:
    ap = argparse.ArgumentParser(prog="wiki", description=__doc__,
                                 formatter_class=argparse.RawDescriptionHelpFormatter)
    sub = ap.add_subparsers(dest="cmd", required=True)
    sub.add_parser("lint", help="check every invariant").set_defaults(fn=cmd_lint)
    sub.add_parser("status", help="what needs recompiling").set_defaults(fn=cmd_status)
    sub.add_parser("index", help="regenerate wiki/index.md").set_defaults(fn=cmd_index)
    sub.add_parser("graph", help="emit link graph as mermaid").set_defaults(fn=cmd_graph)
    sub.add_parser("stats", help="counts and coverage").set_defaults(fn=cmd_stats)
    n = sub.add_parser("new", help="scaffold a page")
    n.add_argument("slug", help="e.g. concepts/knowledge-decay")
    n.add_argument("--title")
    n.add_argument("--type", default="concept", choices=sorted(PAGE_TYPES))
    n.add_argument("--status", default="provisional", choices=sorted(STATUSES))
    n.set_defaults(fn=cmd_new)
    args = ap.parse_args()
    return args.fn(args)


if __name__ == "__main__":
    sys.exit(main())

File: app/lib/wiki.js
// wiki.js — the compiled-wiki model, in the browser.
//
// A port of tools/wiki.py: same frontmatter subset, same markdown subset, same
// invariants. It has to live here because the app compiles NEW pages client-side
// during ingest, and those pages must be rendered and linted before they exist
// on disk.

export const PAGE_TYPES = ['concept', 'entity', 'comparison', 'question', 'overview'];
export const STATUSES = ['established', 'contested', 'provisional', 'superseded'];
export const REQUIRED = ['title', 'type', 'status', 'updated', 'sources'];
const INFRA = new Set(['index', 'log', 'synthesis']);
const SPECIAL_CITES = new Set(['synthesis']);

// ---------------------------------------------------------------- frontmatter

export function parseFrontmatter(text) {
  if (!text.startsWith('---')) return { meta: {}, body: text };
  const end = text.indexOf('\n---', 3);
  if (end === -1) return { meta: {}, body: text };
  const block = text.slice(3, end).replace(/^\n+|\n+$/g, '');
  const body = text.slice(end + 4).replace(/^\n+/, '');

  const meta = {};
  let key = null;
  for (const line of block.split('\n')) {
    if (!line.trim() || line.trimStart().startsWith('#')) continue;
    if (/^[ \t]/.test(line) && line.trimStart().startsWith('- ') && key) {
      if (!Array.isArray(meta[key])) meta[key] = [];
      meta[key].push(line.trimStart().slice(2).trim().replace(/^["']|["']$/g, ''));
      continue;
    }
    const i = line.indexOf(':');
    if (i === -1) continue;
    key = line.slice(0, i).trim();
    let val = line.slice(i + 1).trim();
    if (val.startsWith('[') && val.endsWith(']')) {
      meta[key] = val.slice(1, -1).split(',')
        .map(s => s.trim().replace(/^["']|["']$/g, '')).filter(Boolean);
    } else {
      val = val.replace(/^["']|["']$/g, '');
      meta[key] = val || [];
    }
  }
  return { meta, body };
}

export function serialize(page) {
  const m = page.meta;
  const srcs = (m.sources || []).map(s => `  - ${s}`).join('\n');
  return `---\ntitle: ${m.title}\ntype: ${m.type}\nstatus: ${m.status}\n` +
    `updated: ${m.updated}\nsources:${srcs ? '\n' + srcs : ' []'}\n---\n\n` +
    page.body.replace(/^\n+/, '');
}

export function stripCode(body) {
  const out = [];
  let fenced = false;
  for (const line of body.split('\n')) {
    if (/^\s*```/.test(line)) { fenced = !fenced; continue; }
    if (!fenced) out.push(line);
  }
  return out.join('\n');
}

export const LINK_RE = /\[\[([^\]|]+?)(?:\|([^\]]*))?\]\]/g;
export const CITE_RE = /\[\^([A-Za-z0-9._-]+)\]/g;
const FOOTDEF_RE = /^\[\^([A-Za-z0-9._-]+)\]:\s*(.*)$/;

export function linksOf(body) {
  return [...stripCode(body).matchAll(LINK_RE)].map(m => m[1].trim().replace(/^\//, ''));
}

export function citesOf(body) {
  const out = [];
  for (const line of stripCode(body).split('\n')) {
    if (FOOTDEF_RE.test(line.trim())) continue;
    for (const m of line.matchAll(CITE_RE)) out.push(m[1]);
  }
  return out;
}

export function asDate(v) {
  if (typeof v !== 'string') return null;
  const m = v.trim().match(/^(\d{4})(?:-(\d{2}))?(?:-(\d{2}))?$/);
  if (!m) return null;
  return new Date(Date.UTC(+m[1], (+m[2] || 1) - 1, +m[3] || 1));
}

// ----------------------------------------------------------------------- lint
// Same invariants as tools/wiki.py lint. Consistency only — never truth.

export function lint(pages, sources) {
  const errors = [], warnings = [];
  const bySlug = new Map(pages.map(p => [p.slug, p]));
  const bySid = new Map(sources.map(s => [s.id, s]));
  const inbound = new Map();
  const err = (w, m) => errors.push(`${w}: ${m}`);
  const warn = (w, m) => warnings.push(`${w}: ${m}`);

  for (const pg of pages) {
    const where = pg.file || `wiki/${pg.slug}.md`;
    const meta = pg.meta || {};
    const isInfra = INFRA.has(pg.slug);

    for (const f of REQUIRED) {
      const v = meta[f];
      if (v === undefined || v === '') err(where, `frontmatter missing '${f}'`);
    }
    if (meta.type && !PAGE_TYPES.includes(meta.type))
      err(where, `type '${meta.type}' not in ${PAGE_TYPES.join(', ')}`);
    if (meta.status && !STATUSES.includes(meta.status))
      err(where, `status '${meta.status}' not in ${STATUSES.join(', ')}`);
    if (meta.updated && !asDate(meta.updated))
      err(where, `updated '${meta.updated}' is not a date`);

    const declared = Array.isArray(meta.sources) ? meta.sources
      : (meta.sources ? [meta.sources] : []);
    const links = linksOf(pg.body);
    const cites = citesOf(pg.body);

    for (const l of links) {
      if (!bySlug.has(l)) err(where, `broken link [[${l}]]`);
      else {
        if (!inbound.has(l)) inbound.set(l, new Set());
        inbound.get(l).add(pg.slug);
      }
    }

    for (const c of new Set(cites)) {
      if (SPECIAL_CITES.has(c)) continue;
      if (!bySid.has(c)) err(where, `citation [^${c}] has no source in raw/`);
      else if (!declared.includes(c))
        err(where, `cites [^${c}] but omits it from frontmatter sources`);
    }
    for (const s of declared) {
      if (!bySid.has(s)) err(where, `frontmatter lists unknown source '${s}'`);
      else if (!cites.includes(s) && !isInfra)
        warn(where, `declares source '${s}' but never cites it`);
    }

    const clean = stripCode(pg.body);
    if (meta.status === 'contested' && !/^##+\s*Contradictions/im.test(clean))
      err(where, "status 'contested' requires a '## Contradictions' section");
    if (meta.status === 'superseded' && links.length === 0)
      err(where, "status 'superseded' must link forward to its replacement");

    const pdate = asDate(meta.updated);
    if (pdate) {
      for (const s of declared) {
        const src = bySid.get(s);
        const sdate = src ? asDate(src.added) : null;
        if (sdate && sdate > pdate)
          err(where, `stale: source '${s}' added ${src.added} but page compiled ` +
            `${meta.updated} — recompile`);
      }
    }

    if (!isInfra) {
      for (const para of clean.split(/\n\s*\n/)) {
        const p = para.trim();
        if (!p || /^[#>|\-*[]/.test(p) || /^\d+\./.test(p)) continue;
        if (p.split(/\s+/).length < 25) continue;
        if (!/\[\^[A-Za-z0-9._-]+\]/.test(p))
          warn(where, `uncited assertion: "${p.replace(/\s+/g, ' ').slice(0, 60)}…"`);
      }
    }
  }

  for (const pg of pages) {
    if (INFRA.has(pg.slug)) continue;
    const where = pg.file || `wiki/${pg.slug}.md`;
    const ins = inbound.get(pg.slug);
    const outs = linksOf(pg.body).length;
    if (!ins || ins.size === 0) warn(where, 'orphan — no page links here');
    else if (outs < 3) warn(where, `only ${outs} outbound links (aim for 3+)`);
  }

  const cited = new Set(pages.flatMap(p => citesOf(p.body)));
  for (const s of sources) {
    if (!cited.has(s.id))
      err(s.file || `raw/${s.id}.md`,
        `source '${s.id}' is in raw/ but no wiki page cites it — not ingested`);
  }

  return { errors, warnings, inbound };
}

export function staleness(pages, sources) {
  const bySid = new Map(sources.map(s => [s.id, s]));
  const stale = [], uningested = [];
  const cited = new Set(pages.flatMap(p => citesOf(p.body)));
  for (const pg of pages) {
    const pdate = asDate(pg.meta?.updated);
    const declared = Array.isArray(pg.meta?.sources) ? pg.meta.sources : [];
    for (const s of declared) {
      const src = bySid.get(s);
      const sdate = src ? asDate(src.added) : null;
      if (pdate && sdate && sdate > pdate)
        stale.push({ slug: pg.slug, source: s, page: pg.meta.updated, added: src.added });
    }
  }
  for (const s of sources) if (!cited.has(s.id)) uningested.push(s.id);
  return { stale, uningested };
}

// ------------------------------------------------------------------- markdown

const esc = s => s.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');

// Private-use sentinels: safe to round-trip through the inline formatters
// without colliding with real page text.
const OPEN = '\uE000', CLOSE = '\uE001';

function inline(text, ctx) {
  let out = esc(text);
  const stash = [];
  out = out.replace(/`([^`]+)`/g, (_, c) => {
    stash.push(`<code>${c}</code>`);
    return OPEN + (stash.length - 1) + CLOSE;
  });

  out = out.replace(LINK_RE, (_, target, label) => {
    const t = target.trim().replace(/^\//, '');
    const pg = ctx.pages?.get(t);
    if (pg) {
      const text2 = (label || '').trim() || pg.meta.title || t;
      return `<a href="#/page/${encodeURIComponent(t)}" class="s-${pg.meta.status}">` +
        `${esc(text2)}</a>`;
    }
    return `<span class="broken" title="broken link — this is a lint error">` +
      `[[${esc(t)}]]</span>`;
  });

  out = out.replace(/\[([^\]]+)\]\((https?:\/\/[^)]+)\)/g,
    (_, l, h) => `<a href="${h}" target="_blank" rel="noopener">${l}</a>`);

  out = out.replace(CITE_RE, (_, id) =>
    `<a href="#/source/${encodeURIComponent(id)}" class="cite"><sup>${esc(id)}</sup></a>`);

  out = out.replace(/\*\*([^*]+)\*\*/g, '<strong>$1</strong>');
  out = out.replace(/(?<![*\w])\*([^*\n]+)\*(?!\*)/g, '<em>$1</em>');

  out = out.replace(new RegExp(OPEN + '(\\d+)' + CLOSE, 'g'), (_, i) => stash[+i]);
  return out;
}

export function renderMarkdown(body, ctx = {}) {
  const lines = body.split('\n');
  const out = [];
  const notes = {};
  let i = 0, para = [], listKind = null;

  const flush = () => {
    if (para.length) { out.push(`<p>${inline(para.join(' '), ctx)}</p>`); para = []; }
  };
  const closeList = () => { if (listKind) { out.push(`</${listKind}>`); listKind = null; } };

  while (i < lines.length) {
    const line = lines[i];

    if (/^\s*```/.test(line)) {
      flush(); closeList();
      i++;
      const buf = [];
      while (i < lines.length && !/^\s*```/.test(lines[i])) buf.push(lines[i++]);
      i++;
      out.push(`<pre><code>${esc(buf.join('\n'))}</code></pre>`);
      continue;
    }

    const fd = line.trim().match(FOOTDEF_RE);
    if (fd) { flush(); closeList(); notes[fd[1]] = fd[2]; i++; continue; }

    if (line.trim().startsWith('|') && i + 1 < lines.length &&
        /^\s*\|[\s:|-]+\|\s*$/.test(lines[i + 1])) {
      flush(); closeList();
      const head = line.trim().replace(/^\||\|$/g, '').split('|').map(c => c.trim());
      i += 2;
      const rows = [];
      while (i < lines.length && lines[i].trim().startsWith('|')) {
        rows.push(lines[i].trim().replace(/^\||\|$/g, '').split('|').map(c => c.trim()));
        i++;
      }
      out.push('<div class="tw"><table><thead><tr>' +
        head.map(c => `<th>${inline(c, ctx)}</th>`).join('') + '</tr></thead><tbody>' +
        rows.map(r => '<tr>' + r.map(c => `<td>${inline(c, ctx)}</td>`).join('') +
          '</tr>').join('') + '</tbody></table></div>');
      continue;
    }

    const h = line.match(/^(#{1,6})\s+(.*)$/);
    if (h) {
      flush(); closeList();
      out.push(`<h${h[1].length}>${inline(h[2], ctx)}</h${h[1].length}>`);
      i++; continue;
    }

    if (line.trim().startsWith('>')) {
      flush(); closeList();
      const buf = [];
      while (i < lines.length && lines[i].trim().startsWith('>')) {
        buf.push(lines[i].trim().replace(/^>\s?/, '')); i++;
      }
      out.push(`<blockquote><p>${inline(buf.join(' '), ctx)}</p></blockquote>`);
      continue;
    }

    const ul = line.match(/^\s*[-*]\s+(.*)$/);
    const ol = line.match(/^\s*\d+\.\s+(.*)$/);
    if ((ul || ol) && line.trim()) {
      const kind = ul ? 'ul' : 'ol';
      flush();
      if (listKind !== kind) { closeList(); out.push(`<${kind}>`); listKind = kind; }
      const item = [(ul || ol)[1]];
      i++;
      while (i < lines.length && lines[i].trim() && /^[ \t]/.test(lines[i]) &&
             !/^\s*[-*]\s/.test(lines[i]) && !/^\s*\d+\.\s/.test(lines[i]) &&
             !/^#{1,6}\s/.test(lines[i])) {
        item.push(lines[i].trim()); i++;
      }
      out.push(`<li>${inline(item.join(' '), ctx)}</li>`);
      continue;
    }

    if (!line.trim()) { flush(); closeList(); i++; continue; }
    para.push(line.trim());
    i++;
  }
  flush(); closeList();
  return { html: out.join('\n'), notes };
}

// --------------------------------------------------------------------- search

export function search(pages, q) {
  const terms = q.toLowerCase().split(/\s+/).filter(Boolean);
  if (!terms.length) return [];
  const hits = [];
  for (const pg of pages) {
    const title = (pg.meta.title || pg.slug).toLowerCase();
    const body = pg.body.toLowerCase();
    let score = 0;
    for (const t of terms) {
      if (title.includes(t)) score += 10;
      score += Math.min(body.split(t).length - 1, 8);
    }
    if (score > 0) {
      const idx = body.indexOf(terms[0]);
      const snippet = idx >= 0
        ? pg.body.slice(Math.max(0, idx - 70), idx + 130).replace(/\s+/g, ' ')
        : pg.body.slice(0, 180).replace(/\s+/g, ' ');
      hits.push({ page: pg, score, snippet });
    }
  }
  return hits.sort((a, b) => b.score - a.score);
}
